AP_Eigen

dgrat opened the pull request:
Included: examples and a reference implementation on the example of the compass learning algorithm.
Eigen is a slim algebra library and already used by the PX4 flight stack.
The Current Implementation of AP_Math is not interface complete and lacks optimization for ARM processors.
Using Eigen will increase performance. Currently, there is an increase is in binary size of the executable by ~3.5k.
The memory consumption should stay the same. The performance increase is platform dependent.
I measured something between 30 % and 200 % on some i386 cpus, but not ARM yet.

dgrat commented:
@lucasdemarchi 
updated, but I will not take care about likely packaging errors of ubuntu/debian. 
Users should just use a valid newlib installation.

dgrat commented:
@lucasdemarchi 
Looks like I was wrong, the problem seems to be correlated with this one here:
http://lists.freebsd.org/pipermail/freebsd-current/2014-January/048130.html

lucasdemarchi commented:
yeah... very bad compiler version. And this is the "official" compiler version for ARM cortex-m. Could you file a bug against https://launchpad.net/gcc-arm-embedded ?

dgrat commented:
https://bugs.launchpad.net/gcc-arm-embedded/+bug/1511761

dgrat commented:
@lucasdemarchi 
I don't want to complain, but I think the ARM GCC is a pain. 
Even with GCC 4.8.* it was not compiling with bypassing some silly guard flags.
This is also the way the PX4 guys do it currently. I also don't like how parts of the standard library are implemented. It looks utterly cheap, as if no one wanted to do it. 
No template is checked ..

lucasdemarchi commented:
Em 30/10/2015 13:06, "Daniel Frenzel" <notifications@github.com> escreveu:
>
> https://bugs.launchpad.net/gcc-arm-embedded/+bug/1511761


Great, thanks. Let's see how things progress.

lucasdemarchi commented:
On Fri, Oct 30, 2015 at 1:19 PM, Daniel Frenzel <notifications@github.com>
wrote:

> @lucasdemarchi <https://github.com/lucasdemarchi>
> I don't want to complain, but I think the ARM GCC is a pain.
> Even with GCC 4.8.* it was not compiling with bypassing some silly guard
> flags.


Bugs happen. Even on compilers ;-).  If they fix it soon we can probably
switch to the next version. Otherwise we will need to find out another way
to solve the issue.

dgrat commented:
@lucasdemarchi 
One guy from stackoverflow was able to build a lgamma example with this toolchain: http://releases.linaro.org/15.05/components/toolchain/binaries/
You know more about compiler stuff. Is it worth a try?

cmarcelo commented:
Hi Daniel, I looked at your patch and have some comments/questions. First of all, I think is a good idea to use a library.

1) x(), y(), z() are already implemented for Eigen vectors. I couldn't understand why the wrapper classes reimplement them.

2) Avoid mixing new API additions at the same time (patches) of the Eigen addition. We can either add them before (if helps our transition) or later (if we can transition without them). It's OK to be in the same PR if they are necessary for the goal, but do as separate patch(es).

3) It seems to me we could live without having to add a wrapper for Matrix and Vector classes (VectorE2f and friends). The things that wrapping buy us (besides all the boilerplate necessary for constructors and copy) in your current version:

- get_nearest_integral() (new API addition)
- constrain*() (new API addition)
- length(), is_inf(), is_nan(), angle() seem to be part of the existing vector/matrix API.

The new additions are new, so we have the freedom to add them as generic non-member functions.

For those that are part of the API already, I wonder if we would be best served by refactoring the current code to use non-member functions (we can provide specialized versions for each from the outside the class), and then using the typedefs without any extra wrapping.

My major concern here is to maintain the review/boilerplate code. In the ideal scenario, we should be able to simply say

    using Eigen::Vector2f; // ...

in our AP_Math.h header.

4) About directories: AP_Math seem a natural place for those AP specific typedefs and extra math functions. To be consistent with other third-party code, I suggest putting "eigen" itself inside modules/. I don't see a need for AP_Eigen.

dgrat commented:
1) Maybe I overlooked something, but I don't use the predefined Eigen vector typedefs here, mainly because the typedefs for vector types in Eigen were not convenient for adding extended base types like I did. 

2) I will try to separate them more

3 i) I think you may overestimate the overhead and I don't really like the term wrapper anymore. This is not really a wrapper. 
3 ii) I added some standalone template functions in Extensions based on AP_Math. Some of these functions have been used in Vector/Matrix classes because they make using these classes easier (see example). I don't think that C-style functions bring us much benefit here. Why we should implement e.g. another function for converting a vector outside of the class, when it belongs to the class? This will lead to fragmentation of the code.

cmarcelo commented:
Thanks for the answers, @dgrat.

1) Definition in Eigen (end of Matrix.h) seem to make exactly the one in the PR. But I might be missing some issue. What was the inconvenience?

3) Sorry if I wasn't clear. I understand that you are not consuming neither more memory per Matrix/Vector. My point was whether we actually needed to have the types that are defined `IntfImp`. Or if we could simply do something like:

```
    using Eigen::Vector2d;
```

From my understanding, today's reason to have the types defined in `IntfImp` is to provide a few member functions, which could be made non-member.

About the `get_nearest_integral()` and `constrain()` they seem good additions and I think they do improve the code, but they don't need to be member functions.

> Why we should implement e.g. another function for converting a vector outside of the class, when it belongs to the class?

Because the idea here is that we don't have to implement a new class at all, we could simply use the existing classes provided by Eigen, and add extra functions for things that we care about. Instead of writing `vector.get_nearest_integral()` we would be writing `get_nearest_intergral(vector)`. I don't think `get_nearest_integral()` is part of the class.

cmarcelo commented on libraries/AP_Eigen/AP_Eigen.h:
```diff
@@ -0,0 +1,479 @@
+        }
+        
+        // some functions
+        float length() const {
```
Eigen has a `norm()` family of functions that provide this. Maybe we should transition to use that?

dgrat commented:
@cmarcelo:  Common part with some explanation for the current design:
  * Initially, I implemented these classes in the hope that they can be used similar to the old vector and matrix classes. Additionally, I added two operators to make conversions to the AP library possible. I wanted to remove these two operators later when most parts of AP made the transition. 
  *  I understand that many people here think that it might be better to use the pure Eigen typedefs instead of these classes for some reasons (mentioned or unmentioned), but there are some concerns in additions to the point above which made doing it different:
    - Derived classes can be extended, the core library of Eigen not, and in addition, in some parts of the library there could be saved a considerable amount of code by implementing functionality in the Matrix/Vector classes
    - At the moment we don't really know for sure which upcoming features will be implemented or make sense, but maybe there will be demand on the side of the guys working on the filters, or sensors..  
    - These derived classes are 100 % compatible with Eigen and convert into each other (so no one is forced to use them). Everyone is free to use Eigen::VectorXY in parts of the library were the added functions are not required. I don't think it would make the code worse to read or bring any harm.
    - This principle is not unusual. Think about Qt3/4/5, there you also derive from a class (e.g. QTableWidgetItem), add some functionality and use it in the TableWidget maybe next to the Items which are just standard..
    - I follow up on the point above, because I like the Qt library. In Qt common functions are usually implemented as static members (always encapsulated into structures which belong together) 

Comprehension:
    - I think, with this approach the transition is easier, the functionality is encapsulated into the target structures and one is free to add future functionality in distinct places (my classes) spearated from the Eigen core library.   

Q/A:
1) According to: http://eigen.tuxfamily.org/dox/group__matrixtypedefs.html I was not able to find typedefs with fixed size but open template type. However, I wanted to add functions on the level of the template library. 

2) I think norm() is not equal to constraint_value() nor to constraint_angle(). Norm() can be used to make a unit vector. Constraint is useful to constraint each variable to a be within a range (either linear value or angles 180°/360°). It is not dependent on the length of the vector or may even change it. The scope of these functions are different, but I found examples in the AP.

At the end, I will not decide in which direction the decision will go, but I honestly hope at the end AP_Math will be replaced with either a pure template based function collection in which all the macros have been removed, or with extended Eigen types. At the moment I also still think about different implementation approaches. Nevertheless, I already oriented on the std library since cxx11 (e.g. namings, type traits, ..).

cmarcelo commented:
Thanks for the explanations.


> Derived classes can be extended, the core library of Eigen not, and in addition, in some parts of the library there could be saved a considerable amount of code by implementing functionality in the Matrix/Vector classes

Most of the data we care about in Matrix/Vector are accessible from outside, so I'm failing to see why extra functions we provide need to be member functions at all. An example here might be helpful to me.

However, if we still want to have these helpers as member functions, please consider the recommended route from Eigen documentation, in the section "Extending MatrixBase (and other classes)" in http://eigen.tuxfamily.org/dox/TopicCustomizingEigen.html. That would remove the need of creating (and maintaining) new classes, while still adding new member functions.


> 1) According to: http://eigen.tuxfamily.org/dox/group__matrixtypedefs.html I was not able to find typedefs with fixed size but open template type. However, I wanted to add functions on the level of the template library.

You can do things like

    template <typename T>
    double helper(const Eigen::Matrix<T, 2, 1>& v)
    {
        return v.norm();
    }

and be able to use from `Vector2d`, `Vector2i` and etc. Eigen documentation goes in a more detailed variations at http://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html.


> 2) I think norm() is not equal to constraint_value() nor to constraint_angle().

I was refering to the `length()` function in line 159. GitHub rendering might made things a bit confusing.

dgrat commented:
@cmarcelo 
Thanks, I think, I will switch to the helpers if possible and put an update into this thread.

lucasdemarchi commented:
> @cmarcelo 
> Thanks, I think, I will switch to the helpers if possible and put an update into this thread.

ok, I'm marking this with rework-needed.

dgrat commented:
@lucasdemarchi 
Thx, I guess I will require some time.

dgrat commented:
@lucasdemarchi @cmarcelo 
I shifted most of the class dependent functions into extension files for either the Matrix or Base class.
So every derived class in the Eigen core library will have these extensions.

State:
  * Everyone can use the Eigen library and will have access to advanced functions :)
  * There are still type converter classes which I kept outside of the extension file for safety reasons
    * These classes now have a Compat_* prefix and can be used if necessary to convert the ArduPilot objects into Eigen objects, but I advice not to use them whenever possible :D

dgrat commented:
The demo code of the LowPassFIlter is btw a benchmark :D

dgrat commented:
Okay, I think I finished the last bits, which haven't been nice.
  * The extensions should be able to get vectorized by the compiler
  * Added a docu to every function
  * removed a warning
  * Deleted the length() function, because Eigen::norm() can be used
  * Made the Algebra header more type safe
  * Added some additional constrain functions
  * inlined the extensions

dgrat commented:
@lucasdemarchi @tridge 
Compilation on PX4 requires either a downgrade of the library or a recent one with hard float support.
Is it possible to switch over to a HF library? I am not an expert and don't know atm which hardware would be excluded, but it seems atm there are just these two possibilities.

lucasdemarchi commented:
@dgrat the stm32 in pixhawk has FPU... It does single-precision floating point in hardware, i.e. when you use `float`, but not when you use `double`. Where did you get the information that using hard float would fix the build? We are actually already using it.

Not replies yet on the bug report :(. Going back to 4.8 would hurt performance (it was the point to switch to 4.9). Maybe 4.9.2 is an option...

lucasdemarchi commented:
@dgrat what happened to the commit message? It's very hard to read, I think there was cut and past errors in your editor because some paragraphs are duplicated.  It would be nice to be able to split the single-commit in more sensible chunks in order to ease review.

lucasdemarchi commented on .gitmodules:
```diff
@@ -1,15 +1,19 @@
-	path = modules/gbenchmark
-	url = git://github.com/google/benchmark.git
+  path = modules/gbenchmark
+  url = git://github.com/google/benchmark.git
```
None of the changes above should be here.

dgrat replied:
Rebase artifact and I was too lazy to fix it.

lucasdemarchi commented on .gitmodules:
```diff
@@ -1,15 +1,19 @@
-	url = git://github.com/google/benchmark.git
+  path = modules/gbenchmark
+  url = git://github.com/google/benchmark.git
+[submodule "libraries/AP_Eigen/eigen"]
```
All git submodules should be checkout at modules/ directory, not in other parts of the tree.

lucasdemarchi commented on libraries/AP_Eigen/AP_Algebra.h:
```diff
@@ -0,0 +1,165 @@
+#include <float.h>
+
+#undef min
+#undef max
```
leftover from when we needed this?

dgrat replied:
Also a rebase leftover :D

lucasdemarchi commented on libraries/AP_Eigen/AP_Algebra.h:
```diff
@@ -0,0 +1,165 @@
+template <class IntType, class FloatType>
+static inline IntType safeFloatToInt(const FloatType &num) {
+    //check if float fits into integer
+    if(std::numeric_limits<IntType>::digits > std::numeric_limits<FloatType>::digits) {
```
We are trying very hard to keep the coding style consistent in the project. It would be really nice if new code like this would follow it. Here is the page with examples: http://dev.ardupilot.com/wiki/style-guide/

The mistakes I'm seeing throughout the PR are in general related to spaces. It should be:
  - `if (std::...)`, not `if(std::..)`;
  - `for (std::...)`, not `for(std::...)`;
  - `std::numeric_limits<IntType>::digits>`, not `std::numeric_limits<IntType>::digits >`;
  - functions should have their braces in a new line, not in the same line. This applies **only** to functions.
  - function and variable names should not use camelCase, use underscores

dgrat replied:
:) okay, but the last example is a comparison

lucasdemarchi replied:
yeah, right :)

lucasdemarchi commented on libraries/AP_Eigen/AP_Algebra.h:
```diff
@@ -0,0 +1,165 @@
+
+/*
+ * @brief: Safe way to convert floating point numbers to integers
+ * @authors: Daniel Frenzel <dgdanielf@gmail.com>
```
I like to give credit where it's due, but adding it to each function is pointless. What happens when some one else modifies the function?  You could add your copyright to the beginning of the file, because you wrote it all (or a great part of it when people start changing it); but let's keep the code clean: git is awesome to return the authorship of each and every line in the codebase, so no need to add it here.

lucasdemarchi commented on libraries/AP_Eigen/AP_Algebra.h:
```diff
@@ -0,0 +1,165 @@
+    static_assert(std::is_floating_point<FloatOne>::value && std::is_floating_point<FloatTwo>::value, 
+        "ERROR - is_equal(): template parameters not of type float\n");
+    
+    return fabsf(fVal1 - fVal2) < FLT_EPSILON ? true : false; 
```
so... if type is double you still use FLT_EPSILON rather than DBL_EPSILON?

dgrat replied:
very good point. I will make it better

lucasdemarchi commented on libraries/AP_Eigen/AP_Eigen.h:
```diff
@@ -0,0 +1,35 @@
+#define EIGEN_MATRIX_PLUGIN     <AP_Eigen/AP_MatrixAddons.h>
+
+/*
+ * The PX4 stack creates a lot of drama if just using C99 math features :D
```
no need to bash/rant in header files... :(

lucasdemarchi commented on libraries/AP_Eigen/AP_Eigen.h:
```diff
@@ -0,0 +1,35 @@
+/*
+ * The PX4 stack creates a lot of drama if just using C99 math features :D
+ */
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 && CONFIG_HAL_BOARD != HAL_BOARD_LINUX // maybe more?!
```
if CONFIG_HAL_BOARD is HAL_BOARD_PX4 it's always != CONFIG_HAL_LINUX. Maybe you meant an || rather than &&?

lucasdemarchi commented on libraries/AP_Eigen/AP_Eigen.h:
```diff
@@ -0,0 +1,35 @@
+    #pragma GCC diagnostic push
+        #define RAND_MAX __RAND_MAX
+        #pragma GCC diagnostic ignored "-Wshadow"
+        #pragma GCC diagnostic ignored "-Wfloat-equal"
```
wow... if eigen is shadowing or use float-equal, the problem is not in PX4, but rather in eigen itself.

And you should use `#pragma GCC diagnostic push` and `#pragma GCC diagnostic pop`, otherwise you will have undesired effects.

dgrat replied:
I am not sure whether the -Wfloat-equal is necessary still, because I was experimenting a lot and I copied this from the PX4 flight stack. The really important part is the C99 macro bypass. Parts of the ARM NONE EABI are guarded by a macro and in recent versions additionally many functions are not accessible. This concerns me a lot. I think it would be better if we could try out a HF library.

lucasdemarchi replied:
> This concerns me a lot. I think it would be better if we could try out a HF library.

this statement doesn't make much sense to me. What do you mean by HF library?

lucasdemarchi commented on libraries/AP_Eigen/AP_MatrixAddons.h:
```diff
@@ -0,0 +1,136 @@
+EIGEN_STRONG_INLINE static Matrix<IntType, _Rows, _Cols, _Options, _MaxRows, _MaxCols> get_nearest_integral(const Matrix<FloatType, _Rows, _Cols, _Options, _MaxRows, _MaxCols> &other) {
+    if(!std::is_integral<IntType>::value || !std::is_floating_point<FloatType>::value) {
+        printf("Matrix::get_nearest_integral(): template parameters do not make sense\n");
+    }
```
should be using static_assert here

dgrat replied:
Actually, I was also in doubt here. I wanted to give a message that the function is not used as intended, but a compiler error was in my eyes too rude. On the other hand, a printf is slow. For performance reasons I will replace it.

lucasdemarchi replied:
and it could possibly spam the log with things that should rather be caught at compile time.

lucasdemarchi commented on libraries/AP_Eigen/eigen:
```diff
@@ -0,0 +1 @@
+Subproject commit e7850ed81f9c469e02df496ef09ae32ec0379b71
```
like I said above, it should be in modules/eigen directory.

lucasdemarchi commented on mk/board_native.mk:
```diff
@@ -9,10 +9,12 @@ DEFINES         =   -DF_CPU=$(F_CPU)
 DEFINES        +=   -DSKETCH=\"$(SKETCH)\" -DSKETCHNAME="\"$(SKETCH)\"" -DSKETCHBOOK="\"$(SKETCHBOOK)\"" -DAPM_BUILD_DIRECTORY=APM_BUILD_$(SKETCH)
 DEFINES        +=   $(EXTRAFLAGS)
 DEFINES        +=   -DCONFIG_HAL_BOARD=$(HAL_BOARD) -DCONFIG_HAL_BOARD_SUBTYPE=$(HAL_BOARD_SUBTYPE)
-WARNFLAGS       =   -Wformat -Wall -Wshadow -Wpointer-arith -Wcast-align -Wno-unused-parameter -Wno-missing-field-initializers
```
-Wall and -Wshadow should not be removed from the warning flags.

dgrat commented:
@lucasdemarchi 
If make -j4 px4 CONFIG_HAL_BOARD == HAL_BOARD_LINUX is true. 
I really hate this build system :D

dgrat commented:
@lucasdemarchi 
Changed:
- I added fopenmp in the native build section, because I want to use vectorization and loop parallelization
- I added a PX4 compatibility header and hope it fixes the build problems 
- I made the type checks (epsilon, ..) safer
- I fixed the commit description
- I removed the author lines

Not changed:
- I didn't added white spaces

dgrat commented:
@lucasdemarchi 
Small update:
 - followed the coding style guide
 - splitted the implementation from the definition in the compatibility class

Any wishes, what might be added?

dgrat commented:
@lucasdemarchi @cmarcelo 
  - Added a variadic template function to calculate the squared distance with unlimited parameter length
  - FIxed a bug related to function naming
  - Fixed compiler warning for my small example

tridge commented:
Hi Daniel, my biggest concern with integrating Eigen is if we can do it without expanding the amount of flash space we use on PX4 significantly. Do you have any numbers on the firmware text size with/without eigen when built for px4?

dgrat commented:
Hi Andrew,
I think the size comparison is tricky because Eigen is a header only template library.
Each function will only occupy space if it is used somewhere in the code. The size increase will depend on the class. So at the moment, the addition of this module shouldn't influence flash size. At the beginning of this fork there was as a slight increase, but after adding  most of the functions into extension files, there shouldn't. To give you an impression I added somewhere in ArduCopter some instances of Eigen and listed the increase of size.

Best, Daniel

AP_Eigen fork:
   text    data     bss     dec     hex filename
 912204    2840   57744  972788   ed7f4 /home/dgrat/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf

AP_Eigen fork with Instantiation Eigen::Matrix:
make[1]: Leaving directory '/home/dgrat/ardupilot'
   text    data     bss     dec     hex filename
 912252    2840   57744  972836   ed824 /home/dgrat/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf

Current Master:
    text    data     bss     dec     hex filename
 914012    2844   57744  974600   edf08 /home/dgrat/ardu_master/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf

dgrat commented:
@tridge 
Here we have it, the final comparison after rebasing. Eigen will probably add 80 bytes :D 

    Master
      text    data     bss     dec     hex filename
    914012    2844   57744  974600   edf08 px4fmu-v2_APM.build/firmware.elf

    Before rebase
      text    data     bss     dec     hex filename
    912204    2840   57744  972788   ed7f4 px4fmu-v2_APM.build/firmware.elf

    After rebase
      text    data     bss     dec     hex filename
    914092    2844   57744  974680   edf58 px4fmu-v2_APM.build/firmware.elf
    
    After rebase + Instancing: 
	* Eigen::Vector4i bsaraa;
	* Eigen::Vector2f baraa;    
	* Eigen::Matrix3f foo;
	* Eigen::Vector4f bar;
	* Eigen::Matrix3i foobar;
      text    data     bss     dec     hex filename
    913980    2844   57872  974696   edf68 px4fmu-v2_APM.build/firmware.elf

dgrat commented:
Okay the fun is over. Will it get included or not? 
The current math implementation is neither state of the art, nor memory efficient or fast.
It is also not safe. More important than an increase in memory is safety. My template library will get hopefully more safe and is already way more comfortable to use.

cmarcelo commented:
@dgrat @lucasdemarchi 

Added a few comments. I would prefer not creating an `AP_Eigen` library and instead add things to `AP_Math`, but I'm happy to defer to Daniel choice here.

I still don't understand why Compat classes are necessary. It seems to me we should just have converter functions (the "meat" of those classes anyway). Depending on the code, we'll need the other direction of conversion as well.

Other than that, the structure looks good (but I haven't tested it locally). Thanks for looking into the Addon mechanism of Eigen.

I suggest in the future (after this PR lands!), to turn your benchmark (and any tests) into automated ones using the support recently added by Gustavo Sousa.

dgrat commented:
Q/A) 

> Given that you have a "fix PX4" header, defining missing std stuff (like max()), would it make sense to define std::isnan for PX4 boards?

Tried, but not possible because isnan is a macro in c++ 98

> Minor: I couldn't understand this comment and why the file uses ".imp" extension.

Because I outsourced the template implementation into another file and I totally don't want that it will be compiled separately if I would make it *.cpp file

> I still don't understand why those types need to exist. Would functions that just convert from "AP_Math type" to "Eigen type" suffice?

Because object conversions are performed on the level of constructors. 

> Was the removal of -Werror intentional?

Because I -Werror alone is useless

> EIGEN_MATH 

Sorry, this is a left-over. But the file using it is just an example (in example directory) to simplify the compass learning...

cmarcelo commented:
>> I still don't understand why those types need to exist. Would functions that just convert from "AP_Math type" to "Eigen type" suffice?
> Because object or type conversions are best performed on the level of constructors. 

Why? Is there a performance concern involved that functions would make things slower compared to having a new type?

dgrat commented:
You wouldn't be able to use the '=' operator from e.g. AP_Vector to AP_Eigen. 
Actually, I agree with you that we rather should port the library without these classes. 
On the other hand, these classes cause no overhead and are maybe useful for the EEPROM vector types. 
My biggest concern is that no one will use this lib if he cannot make old code run and test it.

cmarcelo commented:
Not having the syntax sugar is a good thing here, if we indeed want to move things from using old types to use Eigen types. Adding a third type might have no binary size overhead, but adds complexity to how we use this: "should I use Eigen type, or the Compat type?", "should I add this function to Compat type?".

How would this third type help with the EEPROM? Couldn't we serialize / de-serialize Eigen types directly (either with functions or addon member functions)?

So far it seems to me this is not required to move forward, so what about the other way around: let's not add the extra type, but only the conversion functions. If they types necessary, then we add them. The focus now would be on figuring out a way to incrementally change the code to use Eigen types.

What do you think?

dgrat commented:
Make a suggestion and illustrate the advantage.

cmarcelo commented:
My english sometimes is confusing, sorry.

Suggestion: to have functions to provide conversion between Eigen types and old types instead of adding Compat types.

Advantage: (1) have less types, reducing room for confusion (since we are adding the whole Eigen family), (2) less code to be maintained (specially boilerplate code to create such types), (3) follow the advice from Eigen document to subclass their types only when strictly necessary.

dgrat commented:
I think I will write converter functions. However, before starting I would like offer my concerns. Such functions are not easy to implement, because addon functions must fit to every possible size of the underlying base class and vectors are treated maybe slightly differently than matrices. I will have to add size checks at run time which will make the conversion slower. At the moment this is solved very efficiently by the fact that conversion is just possible for equal sized objects (Vector3f to Eigen::Vector3f, ..), which I will call specialized classes in the following. I will have to think also about certain scenarios as well:
  - what to do if I want to copy smaller vector into bigger matrices and vice versa
  - what to do if I want to copy matrix into vector and vice versa (Eigen doesn't know about its usage)
As mentioned, we will lose all the comfort, because no copy operator will be there. You are right, in the Eigen guide is written to think carefully when to use derivation and to prefer the the mentions addons. Nevertheless, type conversions are exactly the exception, because of all the side conditions to consider on the level of implementation.
I know it looks a little blown up, because I implemented for three _specialized_ classes type conversions. However, I doubt it will be more complicated to implement three or more _safe_ functions for the same conversion.
On the other hand, it would be really better to have less types and it might be really less confusing.

cmarcelo commented:
Converter functions don't need to be member functions (part of addon). You can make the specialized versions the same way you did in the Compat classes. Their code will look like the conversion code you already have written for those Compat classes.

dgrat commented:
@cmarcelo Noticed that writing converter functions would require inclusion of AP_Math.h or vector2/3.h on the lvel of the base class of Eigen::Matrix, but there are small mistakes in nearly all headers in AP which just work out because the build system is arduino style.
Either wrong includes "float.h", missing semicolons, or missing type definitions which make the compiler crazy. E.g. the inclusion of quaternion.h leads to madness, because the compiler doesn't know about "Vector2l, Vector2 and everything" at this time point.
Inclusion of AP_Math.h guides me to an “Expected unqualified-id before 'namespace'” error because there is probably somehwere a mssing semicolon and in AP_HAL_Namespace.h is also a missing semicolon, which is unrelated to this error, but I wonder how it even compiles. 
I really, don't want to bother with one million little mistakes just to realize three converter functions.
My suggestions is to either remove the compat classes entirely or to wait for someone else to realize these functions. I lack motivation to correct AP_Math, because I want to get rid of it

cmarcelo commented:
If I understand things right, you shouldn't need to make those functions as addons.

cmarcelo commented:
(so they won't be included inside Eigen)

cmarcelo commented:
> I will have to think also about certain scenarios as well:
> 
> what to do if I want to copy smaller vector into bigger matrices and vice versa
> what to do if I want to copy matrix into vector and vice versa (Eigen doesn't know about its usage)

Well, I don't see how things are different at all (except for syntax sugar we can have).

In the long run there'll be only Eigen types, so we don't need a bunch of granular conversion stuff between them and the old types. So having matching conversions is enough, then operate only with Eigen types for example. (I'm assuming when we only use Eigen types, those won't be a problem).

lucasdemarchi commented:
Closing, as there is a new PR, #3380
